packages/integrations/markdoc/components/astroNode.ts

```diff
@@ -0,0 +1,86 @@
+import type { AstroInstance } from 'astro';
+import type { RenderableTreeNode } from '@markdoc/markdoc';
+import Markdoc from '@markdoc/markdoc';
+import { MarkdocError } from '../dist/utils.js';
+import z from 'astro/zod';
+
+export type AstroNode =
+	| string
+	| {
+			component: AstroInstance['default'];
+			props: Record<string, any>;
+			children: AstroNode[];
+	  }
+	| {
+			tag: string;
+			attributes: Record<string, any>;
+			children: AstroNode[];
+	  };
+
+export function createAstroNode(
+	node: RenderableTreeNode,
+	components: Record<string, AstroInstance['default']> = {}
+): AstroNode {
+	components = validateComponents(components);
+
+	if (typeof node === 'string' || typeof node === 'number') {
+		return String(node);
+	} else if (node === null || typeof node !== 'object' || !Markdoc.Tag.isTag(node)) {
+		return '';
+	}
+
+	if (isCapitalized(node.name) && node.name in components) {
+		const component = components[node.name];
+		const props = node.attributes;
+		const children = node.children.map((child) => createAstroNode(child, components));
```

Comment:
Please help me to understand this code:
1. `validateComponents` doesn't have a signature of what's returned; what does return? (I'm not familiar with `zod`)
2. once we have `components =` assigned, we pass it to ALL the children of the node and validate again and again... it seems inefficient. 

	This means that if we have a node with 15 children, we validate the same components 15 times. 

	 Maybe the validation should be moved out of `createAstroNode`, or make sure that we validate them only **ONCE**

Reply:
Good questions!
1. `validateComponents` will return the result of Zod's `parse` function, which will just be an identity function if no transforms are being applied (in this case, there are none). There's arguably no need for the variable assignment because of this, so I'll remove it 👍 
2. You're right, I missed this! We should move the validation step outside of `createAstroNode` to the `Renderer.astro` level before create is called